### DISCUSSION TITLE
Prepare image_files for mulitple channels and seg_files for ImageDataset

### DISCUSSION THREAD

**Discussion by yarkinslav** [Project-MONAI/MONAI#5301] [Q&A] [answered]
Hi, I am trying to set up `monai.data.ImageDataset`, but can't understand how to create `image_files` and `seg_files` dicts for input. By now I have created dict `train_files` for 10 subjects:
```
[{'image': ['/input/sub-8/T1.nii.gz',
   '/input/sub-8/feauture_map0.nii.gz',
   '/input/sub-8/feauture_map1_mni.nii',
   '/input/sub-8/feauture_map2.nii.gz',
   '/input/sub-8/feauture_map3.nii.gz'],
  'mask': '/input/sub-8/sub-8T1_brain-final_mask.nii.gz',
  'seg': '/input//label/8.nii.gz'},
 {'image': ['/input/sub-27/T1.nii.gz',
   '/input/sub-27/feauture_map0.nii.gz',
   '/input/sub-27/feauture_map1_mni.nii',
   '/input/sub-27/feauture_map2.nii.gz',
   '/input/sub-27/feauture_map3.nii.gz'],
  'mask': '/input/sub-27/sub-27T1_brain-final_mask.nii.gz',
  'seg': '/input//label/27.nii.gz'},
 {'image': ['/input/sub-53/T1.nii.gz',
   '/input/sub-53/feauture_map0.nii.gz',
   '/input/sub-53/feauture_map1_mni.nii',
   '/input/sub-53/feauture_map2.nii.gz',
   '/input/sub-53/feauture_map3.nii.gz'],
  'mask': '/input/sub-53/sub-53T1_brain-final_mask.nii.gz',
  'seg': ‘/input/label/53.nii.gz'},
...
```
Then I am create dataset and dataloader with:
``` 
train_ds = monai.data.ImageDataset(train_files, seg_files=train_files['seg'], transform=train_transf)`  with no error.
check_loader = DataLoader(train_ds, batch_size=1, num_workers=0)
```

and have an error:
`RuntimeError: LoadImage cannot find a suitable reader for file: ('image', 'mask', 'seg').`

I assume that dict for `seg_files` should be for each subject, But how to pass multiple channels for image?
Thank you

**Comment by Nic-Ma** [comment, marked as answer]
Hi @relyativist ,

Thanks for your interest here.
The `ImageDataset` supposes to work with array data instead of dict data.
For your case, I would suggest to use `Dataset` and dict based transforms:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/dataset.py#L57
You can refer to the Brats tutorial for more details:
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb

Thanks.

**Reply by yarkinslav** [reply]
@Nic-Ma I assume that I should stick with 'ImageDataset' because I want to create `PatchDataset` from images. Following 
https://docs.monai.io/en/stable/data.html#patchdataset, to create Patch Dataset, I have to assign data as `ImageDataset`.  
Or there are options to Patch Dict dataset, like `monai.data.Dataset`?

**Reply by KumoLiu** [reply]
Hi @relyativist,  first you may need to take the `feature map` out of the `image` and turn it into:
```
data = [
{'image': '/input/sub-8/T1.nii.gz',
  'map0': '/input/sub-8/feauture_map0.nii.gz',
  'map1': '/input/sub-8/feauture_map1_mni.nii',
  'map2':'/input/sub-8/feauture_map2.nii.gz',
  'map3':'/input/sub-8/feauture_map3.nii.gz',
  'mask': '/input/sub-8/sub-8T1_brain-final_mask.nii.gz',
  'seg': '/input//label/8.nii.gz'},
.....
]
```
And then you may try `PatchIterd` and `GridPatchDataset` like this to create Patch Dataset:
```
dataset = Dataset(data=data, transform=pre_transforms)
patch_iter = PatchIterd(keys=["image", "seg"], patch_size=(64, 64), start_pos=(0, 0))
ds = GridPatchDataset(data=dataset, patch_iter=patch_iter, transform=transforms, with_coordinates=True)
```
Hope it can help you! Thanks!

**Reply by KumoLiu** [reply]
Hi @relyativist, if you want to stack all the files in the `image` key, you can just keep your original `train_files`, ignore what I said about taking the `feature map` out of `image`, thanks!

**Reply by yarkinslav** [reply]
@KumoLiu Could you please clarify differences between `transforms` in `GridPatchDataset` and `Dataset`
Thanks

**Reply by KumoLiu** [reply]
Hi @relyativist, `transforms` in `GridPatchDataset ` is a callable data transform that operates on the patches, and `transforms` in `Dataset` operates on the images if you need some preprocess on the images before you create patches. Thanks!
